File: frontend-admin-pro/src/api/subcontract.ts
import { http } from '@/utils/http'
import type { ListResp } from '@/types/api'

export type SubcontractOrderBrief = {
  id: number
  code: string
  status: string
  supplier_id: number
  supplier_name: string | null
  created_at: string
}

export type SubcontractOrderItemOut = {
  id: number
  order_id: number
  sku_id: number
  process_id: number | null
  qty: number
  unit_price: string | null
  sent_qty: number
  received_qty: number
  remark: string | null
  sku_code: string | null
  sku_name: string | null
  process_name: string | null
}

export type SendLogOut = {
  id: number
  order_id: number
  item_id: number
  qty: number
  remark: string | null
  created_at: string
  sku_code: string | null
  sku_name: string | null
}

export type ReceiveLogOut = {
  id: number
  order_id: number
  item_id: number
  qty: number
  remark: string | null
  created_at: string
  sku_code: string | null
  sku_name: string | null
}

export type SubcontractOrderOut = SubcontractOrderBrief & {
  remark: string | null
  created_by: number | null
  updated_at: string
  items: SubcontractOrderItemOut[]
  send_logs: SendLogOut[]
  receive_logs: ReceiveLogOut[]
}

export type SubcontractOrderCreateItemIn = {
  sku_id: number
  process_id?: number | null
  qty: number
  unit_price?: string | null
  remark?: string | null
}

export type SubcontractOrderCreateIn = {
  supplier_id: number
  code: string
  remark?: string | null
  items: SubcontractOrderCreateItemIn[]
}

export type SendLogIn = {
  item_id: number
  qty: number
  remark?: string | null
}

export type ReceiveLogIn = {
  item_id: number
  qty: number
  remark?: string | null
}

export const subcontractApi = {
  listOrders(params: any) {
    return http.request<ListResp<SubcontractOrderBrief>>({
      url: '/admin/subcontract',
      method: 'GET',
      params,
    })
  },
  createOrder(data: SubcontractOrderCreateIn) {
    return http.request<SubcontractOrderOut>({
      url: '/admin/subcontract',
      method: 'POST',
      data,
    })
  },
  getOrder(id: number) {
    return http.request<SubcontractOrderOut>({
      url: `/admin/subcontract/${id}`,
      method: 'GET',
    })
  },
  updateStatus(id: number, status: string) {
    return http.request<SubcontractOrderOut>({
      url: `/admin/subcontract/${id}/status`,
      method: 'PATCH',
      params: { status },
    })
  },
  sendLog(id: number, data: SendLogIn) {
    return http.request<SubcontractOrderOut>({
      url: `/admin/subcontract/${id}/send`,
      method: 'POST',
      data,
    })
  },
  receiveLog(id: number, data: ReceiveLogIn) {
    return http.request<SubcontractOrderOut>({
      url: `/admin/subcontract/${id}/receive`,
      method: 'POST',
      data,
    })
  },
}

File: backend/app/schemas/subcontract.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class SubcontractOrderItemIn(BaseModel):
    sku_id: int
    process_id: int | None = None
    qty: int
    unit_price: Decimal | None = None
    remark: str | None = None


class SubcontractOrderItemOut(BaseModel):
    id: int
    order_id: int
    sku_id: int
    process_id: int | None
    qty: int
    unit_price: Decimal | None
    sent_qty: int
    received_qty: int
    remark: str | None
    sku_code: str | None = None
    sku_name: str | None = None
    process_name: str | None = None


class SubcontractOrderCreate(BaseModel):
    supplier_id: int
    code: str | None = None
    remark: str | None = None
    items: list[SubcontractOrderItemIn]


class SubcontractOrderOut(BaseModel):
    id: int
    supplier_id: int
    code: str
    status: str
    remark: str | None
    created_by: int | None
    created_at: datetime
    updated_at: datetime
    supplier_name: str | None = None
    items: list[SubcontractOrderItemOut] = []


class SubcontractOrderBrief(BaseModel):
    id: int
    code: str
    status: str
    supplier_id: int
    supplier_name: str | None = None
    created_at: datetime


class SendLogIn(BaseModel):
    item_id: int
    qty: int
    remark: str | None = None


class ReceiveLogIn(BaseModel):
    item_id: int
    qty: int
    remark: str | None = None


class SendLogOut(BaseModel):
    id: int
    order_id: int
    item_id: int
    qty: int
    remark: str | None
    sent_by: int | None
    sent_at: datetime


class ReceiveLogOut(BaseModel):
    id: int
    order_id: int
    item_id: int
    qty: int
    remark: str | None
    received_by: int | None
    received_at: datetime
